lib/web/resources.go

```diff
@@ -329,14 +343,189 @@ func listResources(clt resourcesAPIGetter, r *http.Request, resourceKind string)
+				return nil, trace.Wrap(err)
+			}
+
+			var apps types.Apps
+			for _, server := range appServers {
+				apps = append(apps, server.GetApp())
+			}
+
+			cfg.Apps = types.DeduplicateApps(apps)
+
+			return &listResourcesGetResponse{
+				Items: ui.MakeApps(cfg),
+			}, nil
+		}
+		return nil, trace.Wrap(err)
+	}
+
+	appServers, err := types.ResourcesWithLabels(resp.Resources).AsAppServers()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var apps types.Apps
+	for _, server := range appServers {
+		apps = append(apps, server.GetApp())
+	}
+
+	cfg.Apps = apps
+
+	return &listResourcesGetResponse{
+		Items:      ui.MakeApps(cfg),
+		StartKey:   &resp.NextKey,
+		TotalCount: &resp.TotalCount,
+	}, nil
+}
+
+func handleClusterDesktopsGet(clt resourcesAPIGetter, r *http.Request) (*listResourcesGetResponse, error) {
+	resp, err := listResources(clt, r, types.KindWindowsDesktop)
+	if err != nil {
+		if trace.IsNotImplemented(err) {
+			windowsDesktops, err := clt.GetWindowsDesktops(r.Context(), types.WindowsDesktopFilter{})
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			windowsDesktops = types.DeduplicateDesktops(windowsDesktops)
+
+			return &listResourcesGetResponse{
+				Items: ui.MakeDesktops(windowsDesktops),
+			}, nil
+		}
+		return nil, trace.Wrap(err)
+	}
+
+	windowsDesktops, err := types.ResourcesWithLabels(resp.Resources).AsWindowsDesktops()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &listResourcesGetResponse{
+		Items:      ui.MakeDesktops(windowsDesktops),
+		StartKey:   &resp.NextKey,
+		TotalCount: &resp.TotalCount,
+	}, nil
+}
+
+func handleClusterKubesGet(clt resourcesAPIGetter, r *http.Request) (*listResourcesGetResponse, error) {
+	resp, err := listResources(clt, r, types.KindKubernetesCluster)
+	if err != nil {
+		if trace.IsNotImplemented(err) {
+			kubeServices, err := clt.GetKubeServices(r.Context())
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+
+			return &listResourcesGetResponse{
+				Items: ui.MakeLegacyKubeClusters(kubeServices),
+			}, nil
+		}
+		return nil, trace.Wrap(err)
+	}
+
+	clusters, err := types.ResourcesWithLabels(resp.Resources).AsKubeClusters()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &listResourcesGetResponse{
+		Items:      ui.MakeKubeClusters(clusters),
+		StartKey:   &resp.NextKey,
+		TotalCount: &resp.TotalCount,
+	}, nil
+}
+
 type listResourcesGetResponse struct {
 	// Items is a list of resources retrieved.
 	Items interface{} `json:"items"`
 	// StartKey is the position to resume search events.
-	StartKey string `json:"startKey"`
+	// Pointer to use the vallue "nil" to determine
+	// if pagination is supported in the web UI.
+	StartKey *string `json:"startKey"`
 	// TotalCount is the total count of resources available
 	// after filter.
-	TotalCount int `json:"totalCount"`
+	// Pointer to use the vallue "nil" to determine
+	// if pagination is supported in the web UI.
+	TotalCount *int `json:"totalCount"`
```

Comment:
I noticed that on the frontend, we check if pagination is supported by doing this check:

```typescript
paginationUnsupported: json?.startKey === null && json?.totalCount === null
```

Does the explicit check on `null` mean that `undefined` is a valid value for the field?

If not, then what do you think about adding `omitempty` to these two JSON fields? Then the check on the frontend would become

```typescript
paginationUnsupported: !json?.startKey && !json?.totalCount
```

Reply:
actually, that `?` probably confused you and is unnecessary there. These fields as it currently is, would never be `not defined`

So with `omitempty` this current result `{items: [], totalCount: 0, StartKey: ""}` will give me `{items: []}` (same as if not setting the fields either). With this I wouldn't be able to tell if this field is undefined from `no next key or no result` or that it was deliberately not set. With the unset pointer, it gives me this: `{items: [], totalCount: null, startKey: null}`.